### ISSUE TITLE
Tests fail

### ISSUE THREAD

**Issue by szabgab** [SamuelJansen/python_helper#1]
I cloned the repo and tried to run the test using

```
pytest Tests.py
```

and got a bunch of errors:

[out.txt](https://github.com/SamuelJansen/python-helper/files/9740029/out.txt)

Should the tests be executed in a different way?

**Comment by SamuelJansen** [comment]
You actually got only one error

```
assert datetime.datetime.now() == DateTimeHelper.dateTimeNow()
```

I'm not sure why it happened on your machine though
I did added some better logs to verify what's going on. If you pull and rerun the tests on your end, 
we might figure out what the problem is

---

Also, this is a implemented Test environment in itself.
It's a similar architecture to pytests, but with some twics

The following block at `Tests,py`
```python
TestHelper.run(__file__)
```
will run all tests under the package `./python_helper/api/tests` (and also any package named `test` anywhere else, whitch is not the case)

So, basically, you can run the tests in your prompt just executing `Tests.py`
```
python Tests.py
```

I'll add it at the readme

**Comment by szabgab** [comment]
Interesting, why do you prefer this implementation, why not let pytest do it thing?

I tried `python Tests.py` it had a bunch of successes and then the following failure: (after pulling the latest from github)

```
2022-10-09 07:07:15.742257 [ERROR  ] DateTimeHelperTest.dateTime_now failed
Traceback (most recent call last):
  File "/home/gabor/os/python-helper/python_helper/api/src/service/TestHelper.py", line 39, in unitTest
    testReturns[data[1]] = data[0]()
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/EnvironmentAnnotation.py", line 15, in innerResourceInstanceMethod
    raise wraperException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/EnvironmentAnnotation.py", line 10, in innerResourceInstanceMethod
    methodReturn = resourceInstanceMethod(*innerArgs,**innerKwargs)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 57, in innerResourceInstanceMethod
    raise handlerException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 52, in innerResourceInstanceMethod
    handlerReturn = handleAfter(resourceInstanceMethod, callAfter, argsOfCallAfter, kwargsOfCallAfter, returns, methodReturn, inspectGlobals, methodReturnException=methodReturnException, logResult=logResult)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 83, in handleAfter
    raise methodReturnException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 47, in innerResourceInstanceMethod
    methodReturn = resourceInstanceMethod(*innerArgs,**innerKwargs)
  File "/home/gabor/os/python-helper//python_helper/api/test/api/src/service/DateTimeHelperTest.py", line 96, in dateTime_now
    assert datetime.datetime.now() == DateTimeHelper.dateTimeNow(), f'datetime.datetime.now() == DateTimeHelper.dateTimeNow() => {datetime.datetime.now()} == {DateTimeHelper.dateTimeNow()}'
AssertionError: datetime.datetime.now() == DateTimeHelper.dateTimeNow() => 2022-10-09 07:07:15.742111 == 2022-10-09 07:07:15.742118
2022-10-09 07:07:15.743210 [ERROR  ] DateTimeHelperTest failed
Traceback (most recent call last):
  File "/home/gabor/os/python-helper/python_helper/api/src/service/TestHelper.py", line 127, in runModuleTests
    allDidRun, someDidRun, testTime, testReturns = runnableTddModule(testName, testModule, dataList, times, runSpecificTests, testsToRun, allShouldRun, someShouldRun, logResult)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/EnvironmentAnnotation.py", line 15, in innerResourceInstanceMethod
    raise wraperException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/EnvironmentAnnotation.py", line 10, in innerResourceInstanceMethod
    methodReturn = resourceInstanceMethod(*innerArgs,**innerKwargs)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 57, in innerResourceInstanceMethod
    raise handlerException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 52, in innerResourceInstanceMethod
    handlerReturn = handleAfter(resourceInstanceMethod, callAfter, argsOfCallAfter, kwargsOfCallAfter, returns, methodReturn, inspectGlobals, methodReturnException=methodReturnException, logResult=logResult)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 83, in handleAfter
    raise methodReturnException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 47, in innerResourceInstanceMethod
    methodReturn = resourceInstanceMethod(*innerArgs,**innerKwargs)
  File "/home/gabor/os/python-helper/python_helper/api/src/service/TestHelper.py", line 90, in tddModule
    testName, data, testReturns, someDidRun, logResult, discountTimeEnd = runnableUnitTest(testModule, testName, data, testReturns, logResult)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/EnvironmentAnnotation.py", line 15, in innerResourceInstanceMethod
    raise wraperException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/EnvironmentAnnotation.py", line 10, in innerResourceInstanceMethod
    methodReturn = resourceInstanceMethod(*innerArgs,**innerKwargs)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 57, in innerResourceInstanceMethod
    raise handlerException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 52, in innerResourceInstanceMethod
    handlerReturn = handleAfter(resourceInstanceMethod, callAfter, argsOfCallAfter, kwargsOfCallAfter, returns, methodReturn, inspectGlobals, methodReturnException=methodReturnException, logResult=logResult)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 83, in handleAfter
    raise methodReturnException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 47, in innerResourceInstanceMethod
    methodReturn = resourceInstanceMethod(*innerArgs,**innerKwargs)
  File "/home/gabor/os/python-helper/python_helper/api/src/service/TestHelper.py", line 47, in unitTest
    raise unitTestException
  File "/home/gabor/os/python-helper/python_helper/api/src/service/TestHelper.py", line 39, in unitTest
    testReturns[data[1]] = data[0]()
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/EnvironmentAnnotation.py", line 15, in innerResourceInstanceMethod
    raise wraperException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/EnvironmentAnnotation.py", line 10, in innerResourceInstanceMethod
    methodReturn = resourceInstanceMethod(*innerArgs,**innerKwargs)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 57, in innerResourceInstanceMethod
    raise handlerException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 52, in innerResourceInstanceMethod
    handlerReturn = handleAfter(resourceInstanceMethod, callAfter, argsOfCallAfter, kwargsOfCallAfter, returns, methodReturn, inspectGlobals, methodReturnException=methodReturnException, logResult=logResult)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 83, in handleAfter
    raise methodReturnException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 47, in innerResourceInstanceMethod
    methodReturn = resourceInstanceMethod(*innerArgs,**innerKwargs)
  File "/home/gabor/os/python-helper//python_helper/api/test/api/src/service/DateTimeHelperTest.py", line 96, in dateTime_now
    assert datetime.datetime.now() == DateTimeHelper.dateTimeNow(), f'datetime.datetime.now() == DateTimeHelper.dateTimeNow() => {datetime.datetime.now()} == {DateTimeHelper.dateTimeNow()}'
AssertionError: datetime.datetime.now() == DateTimeHelper.dateTimeNow() => 2022-10-09 07:07:15.742111 == 2022-10-09 07:07:15.742118
Traceback (most recent call last):
  File "/home/gabor/os/python-helper/Tests.py", line 2, in <module>
    TestHelper.run(__file__, inspectGlobals=False)
  File "/home/gabor/os/python-helper/python_helper/api/src/service/TestHelper.py", line 214, in run
    allDidRun, didRun, moduleTestTime, testReturns = runModuleTests(testModuleName, runnableTddModule, times, runSpecificTests, testsToRun, logResult, globalsInstance)
  File "/home/gabor/os/python-helper/python_helper/api/src/service/TestHelper.py", line 135, in runModuleTests
    raise methodReturnException
  File "/home/gabor/os/python-helper/python_helper/api/src/service/TestHelper.py", line 127, in runModuleTests
    allDidRun, someDidRun, testTime, testReturns = runnableTddModule(testName, testModule, dataList, times, runSpecificTests, testsToRun, allShouldRun, someShouldRun, logResult)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/EnvironmentAnnotation.py", line 15, in innerResourceInstanceMethod
    raise wraperException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/EnvironmentAnnotation.py", line 10, in innerResourceInstanceMethod
    methodReturn = resourceInstanceMethod(*innerArgs,**innerKwargs)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 57, in innerResourceInstanceMethod
    raise handlerException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 52, in innerResourceInstanceMethod
    handlerReturn = handleAfter(resourceInstanceMethod, callAfter, argsOfCallAfter, kwargsOfCallAfter, returns, methodReturn, inspectGlobals, methodReturnException=methodReturnException, logResult=logResult)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 83, in handleAfter
    raise methodReturnException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 47, in innerResourceInstanceMethod
    methodReturn = resourceInstanceMethod(*innerArgs,**innerKwargs)
  File "/home/gabor/os/python-helper/python_helper/api/src/service/TestHelper.py", line 90, in tddModule
    testName, data, testReturns, someDidRun, logResult, discountTimeEnd = runnableUnitTest(testModule, testName, data, testReturns, logResult)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/EnvironmentAnnotation.py", line 15, in innerResourceInstanceMethod
    raise wraperException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/EnvironmentAnnotation.py", line 10, in innerResourceInstanceMethod
    methodReturn = resourceInstanceMethod(*innerArgs,**innerKwargs)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 57, in innerResourceInstanceMethod
    raise handlerException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 52, in innerResourceInstanceMethod
    handlerReturn = handleAfter(resourceInstanceMethod, callAfter, argsOfCallAfter, kwargsOfCallAfter, returns, methodReturn, inspectGlobals, methodReturnException=methodReturnException, logResult=logResult)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 83, in handleAfter
    raise methodReturnException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 47, in innerResourceInstanceMethod
    methodReturn = resourceInstanceMethod(*innerArgs,**innerKwargs)
  File "/home/gabor/os/python-helper/python_helper/api/src/service/TestHelper.py", line 47, in unitTest
    raise unitTestException
  File "/home/gabor/os/python-helper/python_helper/api/src/service/TestHelper.py", line 39, in unitTest
    testReturns[data[1]] = data[0]()
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/EnvironmentAnnotation.py", line 15, in innerResourceInstanceMethod
    raise wraperException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/EnvironmentAnnotation.py", line 10, in innerResourceInstanceMethod
    methodReturn = resourceInstanceMethod(*innerArgs,**innerKwargs)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 57, in innerResourceInstanceMethod
    raise handlerException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 52, in innerResourceInstanceMethod
    handlerReturn = handleAfter(resourceInstanceMethod, callAfter, argsOfCallAfter, kwargsOfCallAfter, returns, methodReturn, inspectGlobals, methodReturnException=methodReturnException, logResult=logResult)
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 83, in handleAfter
    raise methodReturnException
  File "/home/gabor/os/python-helper/python_helper/api/src/annotation/TestAnnotation.py", line 47, in innerResourceInstanceMethod
    methodReturn = resourceInstanceMethod(*innerArgs,**innerKwargs)
  File "/home/gabor/os/python-helper//python_helper/api/test/api/src/service/DateTimeHelperTest.py", line 96, in dateTime_now
    assert datetime.datetime.now() == DateTimeHelper.dateTimeNow(), f'datetime.datetime.now() == DateTimeHelper.dateTimeNow() => {datetime.datetime.now()} == {DateTimeHelper.dateTimeNow()}'
AssertionError: datetime.datetime.now() == DateTimeHelper.dateTimeNow() => 2022-10-09 07:07:15.742111 == 2022-10-09 07:07:15.742118
```

**Comment by SamuelJansen** [comment]
I use it `@Test()` for experimentation only.
Also, it allows me to set/update/compare/remove environemnt variables meanwhile test is running and get them back to its initial states after the tests

If really needed, I'll switch to pytests; no problem

---
About the broken unit test:

I see the issue. It was a bad implemented unit test.

I fixed it.  It should be working now.

**Comment by szabgab** [comment]
I don't think it is *needed*, I was just wondering and wanted to understand the value in it. What you describe about the environment is usually done in fixtures.

Anyway I wrote a blog post and recorded a video for this issue and the GitHub Actions PR https://code-maven.com/python-helper-let-all-ci-jobs-fail-separately

Thanks for your quick response!
